pkg/mover/intermediate.go

```diff
@@ -0,0 +1,131 @@
+// Finally, images are appended using the go-containerregistry tarball lib
+func saveIntermediateBundle(cd *ChartData, tarFile string, log Logger) error {
+	tmpTarball, err := os.CreateTemp("", "intermediate-bundle-tar-*")
+	if err != nil {
+		return err
+	}
+	tmpTarballFilename := tmpTarball.Name()
+	tfw := wrapAsTarFileWriter(tmpTarball)
+
+	// hints file goes first to be extracted quickly on demand
+	log.Printf("Writing %s...\n", IntermediateBundleHintsFilename)
+	if err := tfw.WriteMemoryFile(IntermediateBundleHintsFilename, cd.rawHints, defaultPerm); err != nil {
+		return fmt.Errorf("failed to write %s: %w", IntermediateBundleHintsFilename, err)
+	}
+
+	log.Printf("Writing Helm Chart files at %s/...\n", cd.chart.Metadata.Name)
+	if err := tarChart(tfw, cd.chart); err != nil {
+		return fmt.Errorf("failed archiving original-chart/: %w", err)
+	}
+
+	if err := packImages(tfw, cd.imageChanges, log); err != nil {
+		return fmt.Errorf("failed archiving images: %w", err)
+	}
+
+	if err := tfw.Close(); err != nil {
+		return fmt.Errorf("failed closing intermediate bundle temp file %s: %w",
+			tmpTarballFilename, err)
+	}
+
+	if err := copyAs(tmpTarballFilename, tarFile); err != nil {
+		return fmt.Errorf("failed copying  %s -> %s: %w", tmpTarballFilename, tarFile, err)
+	}
+	log.Printf("Intermediate bundle complete at %s\n", tarFile)
+	return nil
+}
+
+func copyAs(src, dst string) error {
+	tmpFilename := fmt.Sprintf("%s.tmp", dst)
+	dstFile, err := os.Create(tmpFilename)
+	if err != nil {
+		return fmt.Errorf("failed to create temporary destination file %s: %w", tmpFilename, err)
+	}
+	srcFile, err := os.Open(src)
+	if err != nil {
+		return fmt.Errorf("failed to open source %s: %w", src, err)
+	}
+	if _, err := io.Copy(dstFile, srcFile); err != nil {
+		return fmt.Errorf("failed copying %s -> %s: %w", src, dst, err)
+	}
+	if err := os.Rename(tmpFilename, dst); err != nil {
+		return fmt.Errorf("failed renaming %s -> %s: %w", tmpFilename, dst, err)
+	}
+	return nil
+}
+
+// tarChart tars all files from the original chart into `original-chart/`
+func tarChart(tfw *tarFileWriter, chart *chart.Chart) error {
+	for _, file := range chart.Raw {
+		if err := tfw.WriteMemoryFile(filepath.Join("original-chart", file.Name), file.Data, defaultPerm); err != nil {
+			return fmt.Errorf("failed to write chart's inner file %s: %v", file.Name, err)
+		}
+	}
+	return nil
+}
+
+func packImages(tfw *tarFileWriter, imageChanges []*internal.ImageChange, logger Logger) error {
+	root := "/"
+	fsys := os.DirFS(root)
+	imagesTarFilename, err := tarImages(imageChanges, logger)
+	if err != nil {
+		return fmt.Errorf("failed to pack images: %w", err)
+	}
+	defer os.Remove(imagesTarFilename)
+	fsImagesTarFilename, err := filepath.Rel(root, imagesTarFilename)
+	if err != nil {
+		return fmt.Errorf("failed to get relative path of %s on %s: %w", imagesTarFilename, root, err)
+	}
+	return tfw.WriteFSFile(fsys, fsImagesTarFilename, defaultPerm)
```

Comment:
this might have the wrong name

Reply:
add parameter to indicate a custom name

Reply:
fixed by saving the stream with a name and size